src/System.Management.Automation/engine/runtime/ScriptBlockToPowerShell.cs

```diff
@@ -312,6 +315,28 @@ private ScriptBlockToPowerShellConverter()
             }
         }
 
+        /// <summary>
+        /// Get using values as dictionary for the Foreach-Object cmdlet, and limit the search
+        /// to only the provided scriptblock and no nested scriptblocks.
+        /// </summary>
+        /// <param name = "scriptBlock">Scriptblock to search.</param>
+        /// <param name = "isTrustedInput">True when input is trusted.</param>
+        /// <param name = "context">Execution context.</param>
+        /// <returns>Dictionary of using variable map.</returns>
+        internal static Dictionary<string, object> GetUsingValuesForEachParallel(
+            ScriptBlock scriptBlock,
+            bool isTrustedInput,
+            ExecutionContext context)
+        {
+            return GetUsingValues(
+                body: scriptBlock.Ast,
+                isTrustedInput: isTrustedInput,
+                context: context,
+                variables: null,
+                filterNonUsingVariables: false,
+                searchNestedScriptBlocks: false).Item1;
```

Comment:
This is a breaking change because it will block all nested script blocks even if it's not related to `foreach -parallel`, see the following example
```
$test = 'test1'
1..2 | % -parallel { $v = & { $using:test }; $v + " Yeah" }
```
This script runs fine today, but with this change, it will fail.

Reply:
There is also a problem when you using `start-job` or `invoke-command` with `foreach-object -parallel`,
for example, `Start-Job { $tt = 'test1' 1..2 | % -Parallel { $using:tt } }` still won't work even with this fix,

I think we need to think more about this issue.

Reply:
Hmm, yes.  I think the using variable search must end when entering a second foreach -parallel (nested case only).  Not yet sure how to do this.

Reply:
The same issue potentially happens to a combination use of `Start-Job` and `Invoke-Command`:
```
start-job { $test = '123'; $s = new-session -computer ComputerA; invoke-command { $using:test } -Session $s }
```
So we probably need some more time to discuss the scenarios that makes sense to support and go from there.

Reply:
The first example should not have worked, and I consider it a bug in the original implementation:
```powershell
$test = 'test1'
1..2 | % -parallel { $v = & { $using:test }; $v + " Yeah" }
```
You cannot pass a $using variable into an arbitrary scriptblock, and the error we now get is correct and the same as:
```powershell
$test = 'TestA'
& { $using:test }
InvalidOperation: A Using variable cannot be retrieved. A Using variable can be used only with Invoke-Command, Start-Job, or InlineScript in the script workflow. When it is used with Invoke-Command, the Using variable is valid only if the script block is invoked on a remote computer.
```

The error for Start-Job case has always existed.  Basically you cannot use '$using:' in nested mixed scenarios:
```powershell
start-job { $test = "Hello"; icm . { $using:test } } | wait-job | rcjb
Invoke-Command: The value of the using variable '$using:test' cannot be retrieved because it has not been set in the local session.
```
But we want to change this for nested Foreach-Object -Parallel because it makes sense to support it there, hence this change.

So, unless there are other issues, I feel this change is correct and consistent with other uses of '$using:'

Reply:
Note that the following works for `Start-Job`
```
$test = 'test1'
Start-Job { & { $using:test } }
```
And today, it's consistent to `ForEach-Object -Parallel`.
```
$test = 'test1'
1..2 | % -parallel { & { $using:test } }
```
I believe the consistency is desired.

------------

Even if the above use with `ForEach-Object` is not desired, that behavior will be in PowerShell 7 as we have finalized the `GA-approved` issues, so unless this PR can be approved by the committee and make in the release, it will be a breaking change to take it later.

Reply:
I don't think we should consider this for GA.  But I will look at restricting the variable map search, as I mentioned previously to only affect foreach -parallel nesting, so that we can take this fix later with the most consistency.

Reply:
I can't speak to the technical feasibility, but here's what I think makes for consistent behavior:

I think the guiding principle should be: _only_ nested script-block contexts that _in and of themselves_ require their  own `$using:` references should be exempt from the outer variable-existence check, and that means excluding _only_ the script blocks in the following nested calls: 

* `Invoke-Command -Computer`

* `Start-Job`

* `Start-ThreadJob`

* `ForEach-Object -Parallel`

That means that mere  nested `& { ... }` calls, as in @daxian-dbw's example - as well as nested   `. { ... }` and `pwsh -c { ... }` calls, should continue to be subject to the _caller's_  variable-existence check, and should continue to reference the _caller's_ variable values.